### PR TITLE
feat: implement adaptive word popover positioning based on available screen space

### PR DESCRIPTION
**Pull request by AhmedCodeGuy**
# Summary
Closes: [QF-2866](https://quranfoundation.atlassian.net/browse/QF-2866)

[QF-2866]: https://quranfoundation.atlassian.net/browse/QF-2866?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **New Features**
  * Introduced dynamic popover positioning for word popovers, allowing them to appear on the left, right, or bottom based on available space.
  * Added new styles and CSS variables to support flexible popover placement and sizing.
  * Implemented utility hooks and functions for responsive popover behavior and CSS variable management.

* **Improvements**
  * Enhanced detection of clarification questions to handle case inconsistencies from the API.
  * Improved chapter page synchronization logic for more accurate tracking of last read verses.
  * Updated icons in the notes action for a refreshed visual appearance.

### CODE REVIEW

**Comment by osamasayed** [comment]
bugbot run

**Review by osamasayed — COMMENTED** [review]
## Code Review

**Overall Assessment**: Good implementation that addresses a real UX problem, but has some logical and performance issues that should be addressed.

**✅ Strengths**:
- Clean architecture with well-structured custom hook
- Comprehensive positioning logic with fallback strategies  
- Good documentation with JSDoc comments
- Responsive design with resize handling

**❌ Issues**:

1. **Logic Error** (Line 110):  is inverted. Should be  to show on the side with more space.

2. **Performance Concerns**:
   - No throttling/debouncing for resize events
   - Direct DOM manipulation could cause layout thrashing
   - Multiple DOM queries in positioning calculations

3. **CSS Issues**:
   - Both  and  classes are identical
   -  without proper containment

**📝 Minor Suggestions**:
- Add null checks for  object (SSR compatibility)
- Consider  for DOM measurements
- Add cleanup for CSS custom properties on unmount
- Consider memoizing DOM queries

The core functionality works well, but the logic error and performance issues should be addressed before merging.

**Comment by osamasayed** [comment]
bugbot run

**Comment by osamasayed** [comment]
@coderabbitai review

File: src/hooks/useWindowResizeListener.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    // Call once on mount to initialize
+    callback();
+
+    window.addEventListener('resize', handleResize);
+
+    // Clean up the event listener on unmount
+    return () => {
+      window.removeEventListener('resize', handleResize);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (mzunain) [inline src/hooks/useWindowResizeListener.ts:43]: please specify reason to disable the dependencies

File: src/components/QuranReader/ReadingView/WordPopover/usePopoverPosition.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
+/* eslint-disable react-func/max-lines-per-function */
```
Comment (mzunain) [inline src/components/QuranReader/ReadingView/WordPopover/usePopoverPosition.ts:1]: same here

**Review by mzulqarnain118 — CHANGES_REQUESTED** [review]
## Code Review - PR #2468: Adaptive Word Popover Positioning

### 🔴 Critical Issues

#### 1. **Incorrect Boolean Return in useCssVariables**
**File:** `src/components/QuranReader/ReadingView/WordPopover/useCssVariables.ts`
- The `Object.entries().some()` returns a boolean, but it's being used with the negation operator `!!` which is redundant
- **Issue in:**
  - `QuestionsMenuItem.tsx` line 30
  - `BottomActions.tsx` line 51

```typescript
// Current (Incorrect):
const isClarificationQuestion = !!Object.entries(questionsData?.[verseKey]?.types || {}).some(
  ([type, count]) => type.toUpperCase() === QuestionType.CLARIFICATION && count > 0,
);

// Should be:
const isClarificationQuestion = Object. entries(questionsData?.[verseKey]?.types || {}).some(
  ([type, count]) => type.toUpperCase() === QuestionType.CLARIFICATION && count > 0,
);
```

#### 2. **Performance Issue: Direct DOM Manipulation on documentElement**
**File:** `useCssVariables.ts`
- Setting CSS variables on `document.documentElement` on every state change can cause performance issues
- **Solution:** Use inline styles or a dedicated container element
- Consider using CSS-in-JS or styled-components for better React integration

#### 3. **Memory Leak Potential in useWindowResizeListener**
**File:** `useWindowResizeListener.ts` (line 42-43)
```typescript
// eslint-disable-next-line react-hooks/exhaustive-deps
}, [debouncedCallback, callback, ...dependencies]);
```
- Disabling exhaustive-deps is a red flag
- The `callback` in dependencies can cause infinite loops if it's not memoized by the consumer
- **Solution:** Remove `callback` from dependencies or document that consumers MUST memoize it

#### 4. **Race Condition in usePopoverPosition**
**File:** `usePopoverPosition.ts`
- Multiple state updates in `calculatePopoverPosition` can lead to race conditions
- Should use a single state object or `useReducer` for atomic updates

```typescript
// Current (Problematic):
setPopoverDirection(direction);
setHasEnoughHorizontalSpace(hasEnoughSpace);
setMarginLeft(horizontalMargin);
setMarginRight('0px');
setMarginTop(verticalMargin);

// Better:
setPopoverState({
  direction,
  hasEnoughHorizontalSpace:  hasEnoughSpace,
  marginLeft: horizontalMargin,
  marginRight: '0px',
  marginTop: verticalMargin,
});
```

---

### 🟡 Code Quality Issues

#### 5. **Violation of DRY Principle**
**Files:** `QuestionsMenuItem.tsx` and `BottomActions.tsx`
- The clarification question detection logic is duplicated
- **Solution:** Extract to a utility function:

```typescript
// utils/questions.ts
export const hasClarificationQuestion = (
  questionsData: QuestionsData,
  verseKey: string
): boolean => {
  return Object.entries(questionsData?.[verseKey]?. types || {}).some(
    ([type, count]) => type.toUpperCase() === QuestionType.CLARIFICATION && count > 0,
  );
};
```

#### 6. **Disabled ESLint Rule Without Justification**
**File:** `usePopoverPosition.ts` (line 1)
```typescript
/* eslint-disable react-func/max-lines-per-function */
```
- Hook has 178 lines, violating SRP (Single Responsibility Principle)
- **Solution:** Extract positioning calculation logic into separate utility functions

#### 7. **Complex Calculation Logic Lacks Inline Comments**
**File:** `popoverPositionUtils.ts` - `calculateHorizontalMargin` function
- The RTL calculations (lines 72-85) are complex but lack "why" comments
- Especially the line:  `const marginRightValue = safetyMargin + distanceToRightEdge + popoverWidth;`
- Add comments explaining the mathematical reasoning

#### 8. **Inconsistent Return Type Pattern**
**File:** `useCssVariables.ts`
- This hook doesn't return anything (void) but manages side effects
- Consider returning a cleanup function or status indicator for better testability

#### 9. **Missing Error Handling**
**File:** `popoverPositionUtils.ts` - `getPopoverWidthFromCSS` and `getPopoverHeightFromCSS`
- No handling for when CSS variables are undefined or invalid
- Relies on fallback but doesn't log warnings for debugging

---

### 🟢 Best Practices Violations

#### 10. **React Performance: Unnecessary Re-renders**
**File:** `WordPopover/index.tsx`
- The `usePopoverPosition` hook runs calculations on every render when `isMenuOpened` changes
- Consider memoizing the result with `useMemo`

#### 11. **Conditional Rendering Anti-pattern**
**File:** `WordPopover/index.tsx` (lines 120-122)
```typescript
{isShareModalOpen && (
  <ShareQuranModal isOpen={isShareModalOpen} onClose={onCloseShareModal} verse={word. verse} />
)}
```
- This change is unnecessary - the modal's `isOpen` prop already handles visibility
- **Revert to original:** Just use `<ShareQuranModal isOpen={isShareModalOpen} ...  />`
- Conditional rendering adds extra mount/unmount cycles

#### 12. **SCSS Redundancy**
**File:** `WordPopover. module.scss`
- `.leftSidePopover` and `.rightSidePopover` have identical properties
- **Solution:** Use a single class with modifiers or CSS variables

```scss
.sidePopover {
  margin-left: var(--popover-margin-left);
  margin-right: var(--popover-margin-right, 0);
  margin-top: var(--popover-margin-top, 0);
  position: absolute;
}
```

#### 13. **Missing TypeScript Strict Checks**
**File:** `popoverPositionUtils.ts`
- Parameters like `direction` are optional with defaults, but the function logic doesn't validate edge cases
- What if `wordLeftPx` or `wordRightPx` are negative? Add runtime guards

---

### 🔵 Testing & Documentation

#### 14. **No Unit Tests**
- New utility functions in `popoverPositionUtils.ts` have no corresponding tests
- Complex calculations MUST have test coverage
- **Required:** Add tests for: 
  - `shouldShowOnRight`
  - `checkHorizontalSpace`
  - `calculateHorizontalMargin` (especially RTL cases)
  - `calculateVerticalMargin`

#### 15. **Missing JSDoc for Complex Functions**
- While some functions have JSDoc, the most complex one (`calculateHorizontalMargin`) lacks detailed parameter explanations
- Document the expected ranges for pixel values

---

### 🟣 Security & Edge Cases

#### 16. **Potential Division by Zero**
**File:** `popoverPositionUtils.ts`
- No validation that `window.innerWidth` or `window.innerHeight` are positive
- Add guards for SSR environments where `window` might be undefined

```typescript
export const shouldShowOnRight = (wordLeftPx: number, wordRightPx:  number): boolean => {
  if (typeof window === 'undefined') return false;
  const spaceOnLeft = window.innerWidth - wordLeftPx;
  const spaceOnRight = wordRightPx;
  return spaceOnRight > spaceOnLeft;
};
```

#### 17. **Missing Null Checks**
**File:** `usePopoverPosition.ts` (line 67)
```typescript
const pageElement = wordElement. closest('[id^="page-"]');
```
- If `pageElement` is null and `containerElement` is also null, function returns null
- But callers don't check for null return - will cause runtime errors

---

### 📋 Summary

**Must Fix Before Merge:**
1. Remove redundant `!!` operators (Issues #1)
2. Fix performance issue with CSS variable manipulation (#2)
3. Address race condition with multiple setState calls (#4)
4. Extract duplicated clarification question logic (#5)
5. Add null/undefined guards for SSR (#16)
6. Add unit tests for utility functions (#14)
7. Revert unnecessary conditional rendering change (#11)

**Should Fix:**
- Refactor `usePopoverPosition` to reduce complexity (#6)
- Consolidate SCSS classes (#12)
- Fix memory leak potential (#3)
- Add inline comments for RTL calculations (#7)

**Code Quality Score:** 6.5/10
- Logic is sound, but implementation needs refinement
- Good structure with separated utilities, but execution has issues
- Missing tests is a blocker for production

---

### ✅ What's Good

- ✅ Good separation of concerns (hooks, utils, components)
- ✅ RTL support is considered
- ✅ Proper use of `useLayoutEffect` to prevent visual flickering
- ✅ CSS variables approach is flexible
- ✅ JSDoc documentation for most functions
- ✅ Debounced resize listener for performance

---

**Recommendation:** Request changes - critical issues must be addressed before merging. 

cc: @AhmedCodeGuy
